File: src/main/java/net/bjoernpetersen/volctl/VolumeControl.java
package net.bjoernpetersen.volctl;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.nio.channels.Channels;
import java.nio.channels.FileChannel;
import java.nio.channels.FileLock;
import java.nio.channels.ReadableByteChannel;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;

import static net.bjoernpetersen.volctl.StringUtils.substringAfterLast;
import static net.bjoernpetersen.volctl.StringUtils.substringBeforeLast;

/**
 * Allows master system audio volume access and control.
 * <p>
 * This implementation uses JNI to perform the required native system calls. To do that it has to
 * store an included library file outside of the .jar-file. The location and name of that file can
 * be customized using the constructor parameters.
 *
 * <h2>Usage with multiple class loaders</h2>
 * <p>
 * If you plan to use this class from multiple class loaders, you'll have to set
 * {@code supportMultipleClassLoaders} to {@code true}. If you do that,
 * <b>every instance will export its own library file</b>. These files won't have a fully
 * predictable filename and cannot be deleted by the same JVM that has loaded them.
 * This workaround is necessary because only one {@link ClassLoader} is allowed to load a
 * library file.
 * <p>
 * <b>Note:</b> You may use the {@link #newInstanceWithClassLoaderSupport} method if you
 * want to use the default location and name, but enable multiple class loader support.
 */
@SuppressWarnings({"unused", "WeakerAccess"})
public class VolumeControl {
    /**
     * The minimum value the volume may have.
     */
    public static final int MIN_VOLUME = 0;
    /**
     * The maximum value the volume may have.
     */
    public static final int MAX_VOLUME = 100;

    private static final String LIB_NAME = "volctl";
    private static final String TMP_DIR_PROPERTY_NAME = "java.io.tmpdir";

    /**
     * Constructor.
     *
     * @param dllLocation                 the directory to store the native access library in,
     *                                    defaults to temp dir
     * @param dllName                     the name of the library file without file extension,
     *                                    defaults to "volctl"
     * @param supportMultipleClassLoaders whether to create library files with different names for
     *                                    each new instance (see {@link VolumeControl class docs})
     * @throws IOException if the library file can't be exported or loaded
     */
    public VolumeControl(
        @NotNull Path dllLocation,
        @NotNull String dllName,
        boolean supportMultipleClassLoaders
    ) throws IOException {
        String defaultLibFile = getDefaultLibFileName();
        String extension = substringAfterLast(defaultLibFile, '.');
        final Path path;
        if (supportMultipleClassLoaders) {
            // Every instance gets its own library file
            path = Files.createTempFile(dllLocation, dllName, '.' + extension);
            FileChannel channel = FileChannel.open(path, StandardOpenOption.WRITE);
            writeLibraryFile(channel);
        } else {
            path = dllLocation.resolve(dllName + '.' + extension);
            boolean writeFile = true;
            if (Files.isRegularFile(path)) {
                try {
                    // Try to delete outdated library file
                    Files.delete(path);
                } catch (IOException e) {
                    // Errors are most likely caused by another instance using the existing file,
                    // so we can just ignore it and use the file
                    writeFile = false;
                }
            }

            if (writeFile) {
                FileChannel channel = FileChannel
                    .open(path, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE);
                writeLibraryFile(channel);
            }
        }

        System.load(path.toAbsolutePath().toString());
    }

    /**
     * Same as calling
     * {@link #VolumeControl(Path, String, boolean) VolumeControl(dllLocation, dllName, false)}.
     *
     * @param dllLocation the directory to store the native access library in,
     *                    defaults to temp dir
     * @param dllName     the name of the library file without file extension,
     *                    defaults to "volctl"
     * @throws IOException if the library file can't be exported or loaded
     */
    public VolumeControl(@NotNull Path dllLocation, @NotNull String dllName) throws IOException {
        this(dllLocation, dllName, false);
    }

    /**
     * Same as calling
     * {@link #VolumeControl(Path, String) VolumeControl(dllLocation, getDefaultLibName())}.
     *
     * @param dllLocation the directory to store the native access library in,
     *                    defaults to temp dir
     * @throws IOException if the library file can't be exported or loaded
     */
    public VolumeControl(@NotNull Path dllLocation) throws IOException {
        this(dllLocation, getDefaultLibName());
    }

    /**
     * Same as calling
     * {@link #VolumeControl(Path, String) VolumeControl(getTempDir(), dllName)}.
     *
     * @param dllName the name of the library file without file extension,
     *                defaults to "volctl"
     * @throws IOException if the library file can't be exported or loaded
     */
    public VolumeControl(@NotNull String dllName) throws IOException {
        this(getTempDir(), dllName);
    }

    /**
     * Same as calling
     * {@link #VolumeControl(Path) VolumeControl(getTempDir())}.
     *
     * @throws IOException if the library file can't be exported or loaded
     */
    public VolumeControl() throws IOException {
        this(getTempDir());
    }

    /**
     * Gets the current master audio volume level.
     *
     * @return a value between 0 and 100 (inclusively)
     */
    public int getVolume() {
        return getVolumeNative();
    }

    /**
     * Sets the current master audio volume level.
     *
     * @param value a value between 0 and 100 (inclusively)
     */
    public void setVolume(int value) {
        if (value < MIN_VOLUME)
            throw new IllegalStateException("Value must be positive, was " + value);
        if (value > MAX_VOLUME)
            throw new IllegalStateException("Value must be less than 100, was " + value);
        setVolumeNative(value);
    }

    private native int getVolumeNative();

    private native void setVolumeNative(int value);

    /**
     * Creates a VolumeControl instance with defaults except for the
     * {@code supportMultipleClassLoaders} parameter being true.
     *
     * @return a new VolumeControl instance
     * @throws IOException if the library file couldn't be exported or loaded
     */
    @NotNull
    public static VolumeControl newInstanceWithClassLoaderSupport() throws IOException {
        return new VolumeControl(getTempDir(), getDefaultLibName(), true);
    }

    /**
     * Retrieves the directory for temporary files on the current system.
     *
     * @return the path to the tmp directory
     */
    @NotNull
    public static Path getTempDir() {
        String path = System.getProperty(TMP_DIR_PROPERTY_NAME);
        return Paths.get(path);
    }

    /**
     * @return the default library file name for the current OS, including extension
     */
    @NotNull
    public static String getDefaultLibFileName() {
        return System.mapLibraryName(LIB_NAME);
    }

    /**
     * @return the default library file name for the current OS, excluding extension
     */
    @NotNull
    public static String getDefaultLibName() {
        return substringBeforeLast(getDefaultLibFileName(), '.');
    }

    private static void writeLibraryFile(@NotNull FileChannel channel) throws IOException {
        try {
            FileLock lock = channel.lock();
            try (ReadableByteChannel input = Channels.newChannel(
                VolumeControl.class
                    .getResourceAsStream('/' + getDefaultLibFileName()))) {
                channel.transferFrom(input, 0, Long.MAX_VALUE);
            } finally {
                lock.release();
            }
        } finally {
            channel.close();
        }
    }
}

File: src/main/kotlin/net/bjoernpetersen/volctl/VolumeControl.kt
package net.bjoernpetersen.volctl

import net.bjoernpetersen.volctl.VolumeControl.Companion.newInstanceWithClassLoaderSupport
import java.io.IOException
import java.nio.channels.Channels
import java.nio.channels.FileChannel
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardOpenOption

/**
 * Allows master system audio volume access and control.
 *
 * This implementation uses JNI to perform the required native system calls. To do that it has to
 * store an included library file outside of the .jar-file. The location and name of that file can
 * be customized using the constructor parameters.
 *
 * ## Usage with multiple class loaders
 *
 * If you plan to use this class from multiple class loaders, you'll have to set
 * `supportMultipleClassLoaders` to `true`. If you do that, **every instance will export its own
 * library file**. These files won't have a fully predictable filename and cannot be deleted by
 * the same JVM that has loaded them.
 * This workaround is necessary because only one [ClassLoader] is allowed to load a library file.
 *
 * **Note:** Java-callers may use the [newInstanceWithClassLoaderSupport] method if they want to
 * use the default location and name, but enable multiple class loader support.
 *
 * @param dllLocation the directory to store the native access library in, defaults to temp dir
 * @param dllName the name of the library file without file extension, defaults to "volctl"
 * @param supportMultipleClassLoaders whether to create library files with different names for each
 * new instance (see above)
 */
@Suppress("unused")
class VolumeControl @JvmOverloads constructor(
    dllLocation: Path = getTempDir(),
    dllName: String = getDefaultLibName(),
    supportMultipleClassLoaders: Boolean = false
) {
    init {
        val defaultLibFile = getDefaultLibFileName()
        val extension = defaultLibFile.substringAfterLast('.')
        val path = if (supportMultipleClassLoaders) {
            // Every instance gets its own library file
            Files.createTempFile(dllLocation, dllName, ".$extension").also {
                FileChannel.open(it, StandardOpenOption.WRITE)
                    .writeLibraryFile()
            }
        } else {
            dllLocation.resolve("$dllName.$extension").also {
                try {
                    // Try to delete outdated library file
                    if (Files.isRegularFile(it)) Files.delete(it)
                    FileChannel
                        .open(it, StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE)
                        .writeLibraryFile()
                } catch (e: IOException) {
                    // Errors are most likely caused by another instance using the existing file,
                    // so we can just ignore it and use the file
                }
            }
        }

        System.load(path.toAbsolutePath().toString())
    }

    /**
     * The current master audio volume level. The value is always between 0 and 100 (inclusively).
     */
    var volume: Int
        get() = getVolumeNative()
        set(value) {
            require(value >= MIN_VOLUME) { "Value must be positive, was $value" }
            require(value <= MAX_VOLUME) { "Value must be less than 100, was $value" }
            setVolumeNative(value)
        }

    private external fun getVolumeNative(): Int

    private external fun setVolumeNative(value: Int)

    companion object {
        /**
         * The minimum value the volume may have.
         */
        const val MIN_VOLUME = 0
        /**
         * The maximum value the volume may have.
         */
        const val MAX_VOLUME = 100

        private const val LIB_NAME = "volctl"
        private const val TMP_DIR_PROPERTY_NAME = "java.io.tmpdir"

        /**
         * Java-friendly way to create a VolumeControl instance with defaults except for
         * `supportMultipleClassLoaders` being true.
         *
         * Equivalent to the Kotlin call:
         *
         * ```kotlin
         * VolumeControl(supportMultipleClassLoaders = true)
         * ```
         */
        @JvmStatic
        fun newInstanceWithClassLoaderSupport(): VolumeControl =
            VolumeControl(supportMultipleClassLoaders = true)

        /**
         * Retrieves the directory for temporary files on the current system.
         */
        @JvmStatic
        fun getTempDir(): Path {
            val path = System.getProperty(TMP_DIR_PROPERTY_NAME)
            return Paths.get(path)
        }

        /**
         * @return the default library file name for the current OS, including extension
         */
        @JvmStatic
        fun getDefaultLibFileName(): String = System.mapLibraryName(LIB_NAME)

        /**
         * @return the default library file name for the current OS, excluding extension
         */
        @JvmStatic
        fun getDefaultLibName(): String = getDefaultLibFileName().substringBeforeLast('.')
    }
}

private fun FileChannel.writeLibraryFile() {
    use {
        withLock {
            this::class.java
                .getResourceAsStream("/${VolumeControl.getDefaultLibFileName()}")
                .let { Channels.newChannel(it) }
                .use { input -> transferFrom(input, 0, Long.MAX_VALUE) }
        }
    }
}

private fun FileChannel.withLock(block: (FileChannel) -> Unit) {
    val lock = lock()
    try {
        block(this)
    } finally {
        lock.release()
    }
}
